### PR TITLE
[tests/bug] improve tests and fix a minor bug

### PR DESCRIPTION
**Pull request by felixdittrich92**
This PR:

- updates codecov action to latest
- updates / adds some missing tests (there are still some missing ones but this would go beyond the scope of this PR)
- fixes a minor bug in `merge_strings`  -> `doctr/models/recognition/utils.py`
- add some no-cover for `file_utils.py` and `hub.py`  because we cannot test some cases or it is already tested by third party lib (e.g.: hf hub)

### CODE REVIEW

File: doctr/models/recognition/utils.py [hunk 1]
```diff
@@ -31,7 +31,7 @@ def merge_strings(a: str, b: str, dil_factor: float) -> str:
     """
     seq_len = min(len(a), len(b))
     if seq_len == 0:  # One sequence is empty, return the other
-        return b if len(a) == 0 else b
```
Comment (odulcy-mindee) [inline doctr/models/recognition/utils.py:34]: :joy:

File: tests/common/test_core.py
Comment (odulcy-mindee) [inline tests/common/test_core.py:1]: Why do these tests fail ?
Also, you can just do:
```python
def test_is_tf_available():
    assert not doctr.is_tf_available()

def test_is_torch_available():
    assert doctr.is_torch_available()
```

File: tests/common/test_datasets_utils.py [hunk 2]
```diff
@@ -21,6 +21,11 @@ def test_translate(input_str, vocab, output_str):
     assert out == output_str
 
 
+def test_translate_unknown_vocab():
+    with pytest.raises(KeyError):
+        _ = utils.translate("test", "unknown_vocab")
```
Comment (odulcy-mindee) [inline tests/common/test_datasets_utils.py:26]:
```suggestion
    with pytest.raises(KeyError):
        utils.translate("test", "unknown_vocab")
```
same suggestion for the other contexts

File: tests/tensorflow/test_models_detection_tf.py [hunk 3]
```diff
@@ -39,9 +44,15 @@ def test_detection_models(arch_name, input_shape, output_size, out_prob):
         {CLASS_NAME: np.array([[0.5, 0.5, 1, 1], [0.5, 0.5, 0.8, 0.9]], dtype=np.float32)},
     ]
     # test training model
-    out = model(input_tensor, target, return_model_output=True, return_preds=True, training=True)
+    out = model(
+        input_tensor,
+        target,
+        return_model_output=True,
+        return_preds=True if not train_mode else False,
+        training=True if train_mode else False,
```
Comment (odulcy-mindee) [inline tests/tensorflow/test_models_detection_tf.py:52]:
```suggestion
        return_preds=not train_mode,
        training=train_mode,
```

File: tests/tensorflow/test_models_recognition_tf.py [hunk 4]
```diff
@@ -24,34 +24,49 @@
...
     input_tensor = tf.random.uniform(shape=[batch_size, *input_shape], minval=0, maxval=1)
     target = ["i", "am", "a", "jedi"]
 
-    out = reco_model(input_tensor, target, return_model_output=True, return_preds=True)
+    out = reco_model(
+        input_tensor,
+        target,
+        return_model_output=True,
+        return_preds=True if not train_mode else False,
+        training=True if train_mode else False,
```
Comment (odulcy-mindee) [inline tests/tensorflow/test_models_recognition_tf.py:59]:
```suggestion
        return_preds=not train_mode,
        training=train_mode,
```

**Review by odulcy-mindee — CHANGES_REQUESTED** [review]

File: doctr/models/recognition/utils.py (hunk 1 above)
Comment (felixdittrich92) [reply doctr/models/recognition/utils.py:34]: Added 5min before weekend 😁😅

File: tests/common/test_core.py
Comment (felixdittrich92) [reply tests/common/test_core.py:1]: 1 Test fails only if you develop on doctr (with both backends installed) it is a bit confusing for new contributors... But i agree we should prefer CI (revert follows)

File: tests/tensorflow/test_models_detection_tf.py (hunk 3 above)
Comment (felixdittrich92) [reply tests/tensorflow/test_models_detection_tf.py:52]: 🙈😂 sure i agree

File: tests/tensorflow/test_models_recognition_tf.py (hunk 4 above)
Comment (felixdittrich92) [reply tests/tensorflow/test_models_recognition_tf.py:59]: Same 🙈😂

File: tests/common/test_core.py [hunk 5]
```diff
@@ -1,13 +1,17 @@
+import pytest
+
 import doctr
 
 
 def test_version():
     assert len(doctr.__version__.split(".")) == 3
 
 
+@pytest.mark.skipif(doctr.is_torch_available() and doctr.is_tf_available(), reason="torch and tf are available")
```
Comment (felixdittrich92) [inline tests/common/test_core.py:10]: @odulcy-mindee this is better it skips the tests (`make tests` works now also if both backends installed if you develop on doctr) and does not effect the CI jobs

Comment (odulcy-mindee) [reply tests/common/test_core.py:10]: Oh yeah, that's even better

File: tests/pytorch/test_models_detection_pt.py [hunk 6]
```diff
@@ -14,19 +14,26 @@
...
+        ["db_resnet50", (3, 512, 512), (1, 512, 512), True, True],
+        ["db_resnet50", (3, 512, 512), (1, 512, 512), True, False],
+        ["db_mobilenet_v3_large", (3, 512, 512), (1, 512, 512), True, True],
+        ["db_mobilenet_v3_large", (3, 512, 512), (1, 512, 512), True, False],
+        ["linknet_resnet18", (3, 512, 512), (1, 512, 512), True, True],
+        ["linknet_resnet18", (3, 512, 512), (1, 512, 512), True, False],
+        ["linknet_resnet34", (3, 512, 512), (1, 512, 512), True, True],
+        ["linknet_resnet34", (3, 512, 512), (1, 512, 512), True, False],
+        ["linknet_resnet50", (3, 512, 512), (1, 512, 512), True, True],
+        ["linknet_resnet50", (3, 512, 512), (1, 512, 512), True, False],
```
Comment (odulcy-mindee) [inline tests/pytorch/test_models_detection_pt.py:30]: I just discovered we can stack parametrize parameter [from documentation](https://docs.pytest.org/en/7.2.x/how-to/parametrize.html) (quite useful for `train_mode` argument):
```python
import pytest


@pytest.mark.parametrize("x", [0, 1])
@pytest.mark.parametrize("y", [2, 3])
def test_foo(x, y):
    pass
```

We can keep it like that for this PR, that's fine :+1:

**Review by odulcy-mindee — APPROVED** [review]

File: tests/pytorch/test_models_detection_pt.py (hunk 6 above)
Comment (felixdittrich92) [reply tests/pytorch/test_models_detection_pt.py:30]: Ah nice good to know :+1: 
Yes, I also think that we can slim it down in another PR  :)

**Pull request merged**